{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b8b5a42-3a28-4592-b07b-a552dd66a0e4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# nullpeptide_extraction/extract_peptides.py\n",
    "\n",
    "import re,os,sys,glob\n",
    "import mycustom2\n",
    "from Bio.Seq import Seq \n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "def reader(path):\n",
    "        datafile=open(path,\"r\")\n",
    "        data=datafile.readlines()\n",
    "        datafile.close()\n",
    "        Data=[]\n",
    "        for i in data:\n",
    "                Data+=[i.strip().split('\\t')]\n",
    "        return Data\n",
    "\n",
    "def fast_counter(SequencesL, kmer_lengthL): \n",
    "    Counts = Counter()\n",
    "    for kmer_length in kmer_lengthL:\n",
    "    \tfor seq in SequencesL:\n",
    "        \tfor i in range(0, len(seq)+1-kmer_length):\n",
    "        \t    motif = seq[i:i+kmer_length]\n",
    "        \t    Counts[motif] += 1\n",
    "    return Counts.most_common()\n",
    "\n",
    "\n",
    "filed = \"UP000005640_9606.fasta\"\n",
    "sequenceO = mycustom2.FastaFile(filed)\n",
    "sequencesL = [ i.sequence.upper() for i in sequenceO ]\n",
    "\n",
    "# Peptide lengths to scan\n",
    "kmer_lengthL=range(1,8)\n",
    "Counts=fast_counter(sequencesL,kmer_lengthL)\n",
    "\n",
    "datafile=open(\"one_to_seven_amino_acids_UP000005640_9606\",\"w\")\n",
    "for i in Counts:\n",
    "\tdatafile.write(str(i[0])+'\\t'+str(i[1])+'\\n')\n",
    "datafile.close()\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea27057-7c5f-421a-a04a-1942b0922b12",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# nullpeptide_extraction/generate_all_counts.py\n",
    "\n",
    "import re,os,sys,glob\n",
    "import mycustom \n",
    "from Bio.Seq import Seq \n",
    "from collections import Counter\n",
    "import pickle\n",
    "import io\n",
    "\n",
    "\n",
    "def reader(path):\n",
    "        datafile=open(path,\"r\")\n",
    "        data=datafile.readlines()\n",
    "        datafile.close()\n",
    "        Data=[]\n",
    "        for i in data:\n",
    "                Data+=[i.strip().split('\\t')]\n",
    "        return Data\n",
    "\n",
    "DataL=reader('one_to_seven_amino_acids_UP000005640_9606')\n",
    "DataD = {}\n",
    "for i in DataL:\n",
    "        if \"N\" not in i[0]:\n",
    "                DataD[i[0]]=int(i[1])\n",
    "\n",
    "\n",
    "DataL=reader(\"peptide_kmers_1_7nt\")\n",
    "Data2D={}\n",
    "for k in DataL:\n",
    "        Data2D[k[0]]=0\n",
    "\n",
    "DataTD= { k: DataD.get(k, 0) + Data2D.get(k, 0) for k in set(DataD) | set(Data2D) }\n",
    "\n",
    "datafile=open(\"one_to_seven_amino_acids_UP000005640_9606_total\",\"w\")\n",
    "for i in DataTD.items():\n",
    "        datafile.write(i[0]+'\\t'+str(i[1])+'\\n')\n",
    "datafile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcc7f89f-59ad-472a-bd56-13d23a8267a5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# nullpeptide_extraction/mycustom2.py\n",
    "\n",
    "class GenericFile(object):\n",
    "    def __init__(self, fileNameS):\n",
    "            try:\n",
    "                fileF = open(fileNameS)\n",
    "                self.fileS = fileF.read()\n",
    "                fileF.close()\n",
    "            except:\n",
    "                self.fileS = fileNameS.read()\n",
    "\n",
    "import pdb\n",
    "\n",
    "class FastaFile(GenericFile):\n",
    "    \"\"\"\n",
    "    This class represents a FastaFile.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, fasta, fileName=True):\n",
    "        \"\"\"\n",
    "\n",
    "        :rtype : object\n",
    "        :param fasta: This can be either a file name (string) or a HTML file upload element.\n",
    "        :param fileName: This is a boolean value. If False, then it is the fasta string\n",
    "        :return: does not return anything.\n",
    "        \"\"\"\n",
    "        if fileName == True:\n",
    "            super(FastaFile, self).__init__(fasta)\n",
    "        elif fileName == False:\n",
    "            if \">\" not in fasta:\n",
    "                raise ValueError(\"It is not in Fasta Format\")\n",
    "            self.fileS = fasta\n",
    "        else:\n",
    "            raise Exception\n",
    "\n",
    "        self.sequences = None\n",
    "        self.sequences = self.getSequences()\n",
    "\n",
    "\n",
    "    def __getitem__(self, item):\n",
    "        if self.sequences is None:\n",
    "            self.getSequences()\n",
    "        return self.sequences[item]\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.getSequences())\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"< Fasta file: %s sequences >\" % len(self)\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"< Fasta file: %s sequences >\" % len(self)\n",
    "\n",
    "    def getSequences(self):\n",
    "        if self.sequences is None:\n",
    "            firstSplitL = self.fileS.strip().split(\">\")[1:]\n",
    "            eachByLineL = [i.splitlines() for i in firstSplitL]\n",
    "\n",
    "            sequences = []\n",
    "            for i in eachByLineL:\n",
    "                header = i[0]\n",
    "                sequence = \"\".join(i[1:]).upper()\n",
    "                sequence = sequence.replace(\" \", \"\")\n",
    "                sequence = sequence.replace(\"\\t\", \"\")\n",
    "                if sequence == \"\":\n",
    "                    #raise ValueError(\"There is no sequence !\")\n",
    "\t\t    pass\n",
    "                sequences.append(Sequence(name=header, sequence=sequence))\n",
    "\n",
    "            self.sequences = sequences\n",
    "\n",
    "        return self.sequences\n",
    "\n",
    "    def getMaxLength(self):\n",
    "        \"\"\"\n",
    "        This returns the max length of a sequence in this object\n",
    "        \"\"\"\n",
    "        sequences = self.getSequences()\n",
    "        maxLength = 0\n",
    "        for sequence in sequences:\n",
    "            seqLength = len(sequence)\n",
    "            if seqLength > maxLength:\n",
    "                maxLength = seqLength\n",
    "        return maxLength\n",
    "\n",
    "    def areDuplicatesPresent(self):\n",
    "        sequences = self.getSequences()\n",
    "        sequenceL = [i.sequence for i in sequences]\n",
    "        if len(sequenceL) > len(set(sequenceL)):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "    def getNames(self):\n",
    "\n",
    "        names = [sequence.name for sequence in self.sequences]\n",
    "        return names\n",
    "\n",
    "\n",
    "class Sequence(object):\n",
    "    \"\"\"\n",
    "    This class represents a DNA/RNA sequence.\n",
    "    Each object needs to have a name.\n",
    "    \"\"\"\n",
    "    def __init__(self, name, sequence):\n",
    "        self.name = name\n",
    "        self.sequence = sequence\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"%s : %s\" %(self.name, self.sequence)\n",
    "\n",
    "    def __getitem__(self, item):\n",
    "        return self.sequence[item]\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sequence)\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"%s : %s \" % (self.name, self.sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11e1a88e-a8a8-4673-8e28-8975114aa91d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nullomer_extraction scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "24317031-91c1-4752-96f5-21dbebcb5970",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nullomer_extraction/extract_kmers.py\n",
    "\n",
    "import re,os,sys,glob\n",
    "import mycustom \n",
    "from Bio.Seq import Seq \n",
    "from collections import Counter\n",
    "import pickle\n",
    "\n",
    "\n",
    "def reader(path):\n",
    "        datafile=open(path,\"r\")\n",
    "        data=datafile.readlines()\n",
    "        datafile.close()\n",
    "        Data=[]\n",
    "        for i in data:\n",
    "                Data+=[i.strip().split('\\t')]\n",
    "        return Data\n",
    "\n",
    "def fast_counter(SequencesL,SequencesL_rev_compl, kmer_lengthL): \n",
    "    Counts = Counter()\n",
    "    for kmer_length in kmer_lengthL:\n",
    "    \tfor seq in SequencesL:\n",
    "        \tfor i in range(0, len(seq)+1-kmer_length):\n",
    "        \t    motif = seq[i:i+kmer_length]\n",
    "        \t    Counts[motif] += 1\n",
    "        for seq2 in SequencesL_rev_compl:\n",
    "                for im in range(0, len(seq2)+1-kmer_length):\n",
    "                    motif = seq2[im:im+kmer_length]\n",
    "                    Counts[motif] += 1\n",
    "    return Counts.most_common()\n",
    "\n",
    "\n",
    "filed = \"hg38.fasta\"\n",
    "sequenceO = mycustom.FastaFile(filed)\n",
    "sequencesL = [ i.sequence.upper() for i in sequenceO ]\n",
    "sequencesL_rev_compl = []\n",
    "for i in sequencesL:\n",
    "        seq = Seq(i)\n",
    "        sequencesL_rev_compl+=[str(seq.reverse_complement())]\n",
    "\n",
    "# kmer lengths to scan - here e scan 1bp to 15bp long kmers\n",
    "kmer_lengthL=range(1,16)\n",
    "Counts=fast_counter(sequencesL,sequencesL_rev_compl,kmer_lengthL)\n",
    "\t\n",
    "datafile=open(\"hg38_data\",\"w\")\n",
    "for i in Counts:\n",
    "\tdatafile.write(str(i[0])+'\\t'+str(i[1])+'\\n')\n",
    "datafile.close()\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8082e5fb-1dfe-4811-8572-b588bab1f6a8",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# nullomer_extraction/generate_all_counts.py\n",
    "\n",
    "import re,os,sys,glob\n",
    "import mycustom \n",
    "from Bio.Seq import Seq \n",
    "from collections import Counter\n",
    "import pickle\n",
    "import io\n",
    "\n",
    "\n",
    "def reader(path):\n",
    "        datafile=open(path,\"r\")\n",
    "        data=datafile.readlines()\n",
    "        datafile.close()\n",
    "        Data=[]\n",
    "        for i in data:\n",
    "                Data+=[i.strip().split('\\t')]\n",
    "        return Data\n",
    "\n",
    "DataL=reader('hg38_data')\n",
    "DataD = {}\n",
    "for i in DataL:\n",
    "        if \"N\" not in i[0]:\n",
    "                DataD[i[0]]=int(i[1])\n",
    "\n",
    "\n",
    "DataL=reader(\"kmers_1_15nt\")\n",
    "Data2D={}\n",
    "for k in DataL:\n",
    "        Data2D[k[0]]=0\n",
    "\n",
    "DataTD= { k: DataD.get(k, 0) + Data2D.get(k, 0) for k in set(DataD) | set(Data2D) }\n",
    "\n",
    "datafile=open(\"hg38_data_total\",\"w\")\n",
    "for i in DataTD.items():\n",
    "        datafile.write(i[0]+'\\t'+str(i[1])+'\\n')\n",
    "datafile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd54645-df02-41c4-b2c0-515278427d73",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# nullomer_extraction/mycustom.py\n",
    "\n",
    "class GenericFile(object):\n",
    "    def __init__(self, fileNameS):\n",
    "            try:\n",
    "                fileF = open(fileNameS)\n",
    "                self.fileS = fileF.read()\n",
    "                fileF.close()\n",
    "            except:\n",
    "                self.fileS = fileNameS.read()\n",
    "\n",
    "import pdb\n",
    "\n",
    "class FastaFile(GenericFile):\n",
    "    \"\"\"\n",
    "    This class represents a FastaFile.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, fasta, fileName=True):\n",
    "        \"\"\"\n",
    "\n",
    "        :rtype : object\n",
    "        :param fasta: This can be either a file name (string) or a HTML file upload element.\n",
    "        :param fileName: This is a boolean value. If False, then it is the fasta string\n",
    "        :return: does not return anything.\n",
    "        \"\"\"\n",
    "        if fileName == True:\n",
    "            super(FastaFile, self).__init__(fasta)\n",
    "        elif fileName == False:\n",
    "            if \">\" not in fasta:\n",
    "                raise ValueError(\"It is not in Fasta Format\")\n",
    "            self.fileS = fasta\n",
    "        else:\n",
    "            raise Exception\n",
    "\n",
    "        self.sequences = None\n",
    "        self.sequences = self.getSequences()\n",
    "\n",
    "\n",
    "    def __getitem__(self, item):\n",
    "        if self.sequences is None:\n",
    "            self.getSequences()\n",
    "        return self.sequences[item]\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.getSequences())\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"< Fasta file: %s sequences >\" % len(self)\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"< Fasta file: %s sequences >\" % len(self)\n",
    "\n",
    "    def getSequences(self):\n",
    "        if self.sequences is None:\n",
    "            firstSplitL = self.fileS.strip().split(\">\")[1:]\n",
    "            eachByLineL = [i.splitlines() for i in firstSplitL]\n",
    "\n",
    "            sequences = []\n",
    "            for i in eachByLineL:\n",
    "                header = i[0]\n",
    "                sequence = \"\".join(i[1:]).upper()\n",
    "                sequence = sequence.replace(\" \", \"\")\n",
    "                sequence = sequence.replace(\"\\t\", \"\")\n",
    "                if sequence == \"\":\n",
    "                    raise ValueError(\"There is no sequence !\")\n",
    "                sequences.append(Sequence(name=header, sequence=sequence))\n",
    "\n",
    "            self.sequences = sequences\n",
    "\n",
    "        return self.sequences\n",
    "\n",
    "    def getMaxLength(self):\n",
    "        \"\"\"\n",
    "        This returns the max length of a sequence in this object\n",
    "        \"\"\"\n",
    "        sequences = self.getSequences()\n",
    "        maxLength = 0\n",
    "        for sequence in sequences:\n",
    "            seqLength = len(sequence)\n",
    "            if seqLength > maxLength:\n",
    "                maxLength = seqLength\n",
    "        return maxLength\n",
    "\n",
    "    def areDuplicatesPresent(self):\n",
    "        sequences = self.getSequences()\n",
    "        sequenceL = [i.sequence for i in sequences]\n",
    "        if len(sequenceL) > len(set(sequenceL)):\n",
    "            return True\n",
    "        return False\n",
    "\n",
    "    def getNames(self):\n",
    "\n",
    "        names = [sequence.name for sequence in self.sequences]\n",
    "        return names\n",
    "\n",
    "\n",
    "class Sequence(object):\n",
    "    \"\"\"\n",
    "    This class represents a DNA/RNA sequence.\n",
    "    Each object needs to have a name.\n",
    "    \"\"\"\n",
    "    def __init__(self, name, sequence):\n",
    "        self.name = name\n",
    "        self.sequence = sequence\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"%s : %s\" %(self.name, self.sequence)\n",
    "\n",
    "    def __getitem__(self, item):\n",
    "        return self.sequence[item]\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.sequence)\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"%s : %s \" % (self.name, self.sequence)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea1fa2e2-b344-43f5-aa77-0106901b658a",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# nullomer_extraction/possible_kmers.py import re,os,sys,itertools\n",
    "\n",
    "def kmer_generate(lengthsL):\n",
    "\ttotal_kmers=[]\n",
    "\tbases=['A','T','G','C']\n",
    "\tfor k in lengthsL:\n",
    "\t\ttotal_kmers+=[''.join(p) for p in itertools.product(bases, repeat=k)]\n",
    "\treturn total_kmers\n",
    "\n",
    "KmersL=kmer_generate(range(1,16))\n",
    "\t\t\n",
    "datafile=open(\"kmers_1_15nt\",\"w\")\n",
    "for i in KmersL:\n",
    "        datafile.write(str(i)+'\\n')\n",
    "datafile.close()  \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
